runtime/src/iree/hal/drivers/hip/event_semaphore.c

```diff
@@ -99,484 +132,749 @@ static void iree_hal_hip_semaphore_destroy(
-  }
-
-  // Slow path: acquire a timepoint. This should happen inside of the lock too.
-  // If not locked the semaphore may be signal before acquiring a timepoint.
-  // Then we would miss the signal.
-  iree_status_t status = iree_hal_hip_semaphore_acquire_timepoint_host_wait(
-      semaphore, value, timeout, out_timepoint);
+  iree_status_t status = iree_status_clone(semaphore->failure_status);
+  if (iree_status_is_ok(status)) {
+    iree_hal_hip_util_tree_node_t* node =
+        iree_hal_hip_util_tree_get(&semaphore->event_queue.tree, value);
+
+    if (node == NULL) {
+      status = iree_hal_hip_util_tree_insert(&semaphore->event_queue.tree,
+                                             value, &node);
+      if (iree_status_is_ok(status)) {
+        iree_hal_hip_semaphore_queue_item_t* item =
+            (iree_hal_hip_semaphore_queue_item_t*)
+                iree_hal_hip_util_tree_node_get_value(node);
+        item->cpu_event = NULL;
+        item->work_item = NULL;
+      }
+    }
 
+    if (iree_status_is_ok(status)) {
+      iree_hal_hip_event_t* event =
+          ((iree_hal_hip_semaphore_queue_item_t*)
+               iree_hal_hip_util_tree_node_get_value(node))
+              ->event;
+      if (!event) {
+        do {
+          node = iree_hal_hip_util_tree_node_next(node);
+          if (!node) {
+            status = iree_make_status(IREE_STATUS_FAILED_PRECONDITION,
+                                      "there was no event that could be valid");
+            break;
+          }
+          event = ((iree_hal_hip_semaphore_queue_item_t*)
+                       iree_hal_hip_util_tree_node_get_value(node))
+                      ->event;
+        } while (!event);
+      }
+      if (event) {
+        iree_hal_hip_event_retain(event);
+      }
+      *out_hip_event = event;
+    }
+  }
   iree_slim_mutex_unlock(&semaphore->mutex);
 
   IREE_TRACE_ZONE_END(z0);
+
   return status;
 }
 
-static iree_status_t iree_hal_hip_semaphore_wait(
+iree_status_t iree_hal_hip_semaphore_create_event_and_record_if_necessary(
     iree_hal_semaphore_t* base_semaphore, uint64_t value,
-    iree_timeout_t timeout) {
+    hipStream_t dispatch_stream, iree_hal_hip_event_pool_t* event_pool,
+    iree_hal_hip_event_t** out_hip_event) {
   iree_hal_hip_semaphore_t* semaphore =
       iree_hal_hip_semaphore_cast(base_semaphore);
+  *out_hip_event = NULL;
   IREE_TRACE_ZONE_BEGIN(z0);
-
-  iree_hal_hip_timepoint_t* timepoint;
-  IREE_RETURN_AND_END_ZONE_IF_ERROR(
-      z0, iree_hal_hip_semaphore_try_wait_or_acquire_wait_timepoint(
-              base_semaphore, value, timeout, &timepoint));
-  if (!timepoint) {
-    // We don't need to wait on a timepoint.
-    // The wait condition is satisfied.
-    IREE_TRACE_ZONE_END(z0);
-    return iree_ok_status();
-  }
-
   iree_slim_mutex_lock(&semaphore->mutex);
-  if (semaphore->current_value >= IREE_HAL_SEMAPHORE_FAILURE_VALUE) {
+  if (value <= semaphore->current_visible_value) {
```

Comment:
same as above with tracing only what matters - try functions are usually good candidates for deferred tracing: if the try fails then don't trace and only enter the trace zone once you know it's doing something - otherwise polling can lead to billions of events in large traces when only dozens of actual successes happen. that gives you signal in the trace where seeing the zone means it was necessary as opposed to always seeing the zone but having to guess based on how long it took whether it did anything.